designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,381 @@
+```
+
+### Changing the location of the suppressions file
+
+A new option `--suppressions-location` will be introduced to ESLint CLI. When provided, the suppressions file will be loaded and saved to the provided location. Note that this is a string flag option (value is required).
+
+``` bash
+eslint --suppress-all --suppressions-location /home/user/project/mycache ./src
+```
+
+### Maintaining a lean suppressions file
+
+When working with suppressed violations, it's possible to address a violation without updating the suppressions file. This oversight can allow new violations to go unnoticed. To prevent this, eslint can exit with an error code if there are outdated (unmatched) suppressions.
+
+Consider the following scenario:
+
+* The developer runs `eslint --supress-all ./src` to create the suppressions file.
+* Running `eslint ./src` reports no violations and exits with status 0.
+* After fixing a violation, the suppressions file still contains the now-resolved violation.
+* Running `eslint ./src` again reports no violations but exits with a non-zero status code, indicating the suppressions file needs updating.
+
+To address this, a new option `--prune-suppressions` will be introduced to ESLint. This boolean flag removes resolved violations from the suppressions file without adding new ones, unlike `--suppress-all`.
+
+``` bash
+eslint --prune-suppressions ./src
+eslint --prune-suppressions --suppressions-location /home/user/project/mycache ./src
+```
+
+### Execution details
+
+The suggested solution always compares against the existing suppressions file, typically `.eslint-suppressions.json`, unless `--suppressions-location` is specified.  This makes it easier for existing and new projects to adopt this feature without the need to adjust scripts in `package.json` and CI/CD workflows. 
+
+To perform the comparison, we will go through each result and message from `ESLint.lintFiles`, checking each error `(severity == 2)` against the suppressions file. By design, we ignore warnings since they don't cause eslint to exit with an error code and serve a different purpose. If the file and rule are listed in the suppressions file, we can move the message to `LintResult#suppressedMessages` and ignore the result message.
+
+Here is a high-level overview of the execution flow:
+
+1. **Check for Options**
+  * If both `--suppress-all` and `--suppress-rule` are passed, exit with an error (these options are mutually exclusive).
+  * If either option is passed, update the suppressions file based on the `results`.
+  * If no option is passed, check if the suppressions file exists, considering `--suppressions-location`.
+2. **Match Errors Against Suppressions**
+  * For each error, check if the error and the file are in the suppressions file.
+  * If yes, decrease count, in memory, by 1 and move the message to `LintResult#suppressedMessages` unless count is zero.
+  * If no, keep the error.
```

Comment:
(related to the discussion under https://github.com/eslint/rfcs/pull/119#discussion_r1722953945)

For one file and one rule, I think either all or none of the messages should be moved to suppressedMessages. By this algorithm, when the number of error messages is greater than the number stored in the suppressions file, some would be moved to suppressedMessages while some would stay in messages.